core/trino-main/src/main/java/io/trino/operator/output/PagePartitioner.java

```diff
@@ -107,7 +104,6 @@ public PagePartitioner(
         this.replicatesAnyRow = replicatesAnyRow;
         this.nullChannel = nullChannel.orElse(-1);
         this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
-        this.sourceTypes = sourceTypes.toArray(new Type[0]);
```

Comment:
actually, the `if` there raises questions about potential regressions. 
It can easily produce small pages. For example, if 50% of positions are handled by PostiionsAppenders and the other 50% by BlockBuilders this will cause the average page size to be 50% of the expected page size given enough partitions.
If only 5% of positions go to either one then this will produce a lot of extremely small pages because it will always flush almost empty buffers

Reply:
> It can easily produce small pages. For example, if 50% of positions are handled by PostiionsAppenders and the other 50% by BlockBuilders this will cause the average page size to be 50% of the expected page size given enough partitions.

50% is still big enough too, right?

> If only 5% of positions go to either one then this will produce a lot of extremely small pages because it will always flush almost empty buffers

If you have 5% vs 95%, then you have one small page and one large page. It is fine, see Javadoc for `io.trino.operator.project.MergePages`